Client/Experiments/Messaging/MessagingManager.swift

```diff
@@ -0,0 +1,257 @@
+//        }
+        
+        /// Be sure the count on `triggers` and `message.triggers` are equal.
+        /// If these mismatch, that means a message contains a trigger not in the triggers lookup table.
+        /// JEXLS can only be evaluated on supported triggers. Otherwise, consider the message malformed.
+        if triggers.count != message.trigger.count {
+            return nil
+        }
+        
+        return Message(messageId: messageId,
+                       messageData: message,
+                       action: action,
+                       triggers: triggers,
+                       styleData: style,
+                       metadata: messageStore.getMessageMetadata(messageId: messageId))
+    }
+    
+    /// Returns the next well-formed, non-expired, triggered message in descending priority order.
+    func getNextMessage(for surface: MessageSurfaceId) -> Message? {
+        let feature = FxNimbus.shared.features.messaging.value()
+        let messageStore = MessageStore()
+        guard let helper = messagingHelper.createHelper()else { return nil }
+        
+        /// All these are non-expired, well formed messages for a requested surface.
+        let messages = feature.messages.compactMap { key, messageData -> Message? in
+            if let message = self.createMessage(messageId: key,
+                                                message: messageData,
+                                                lookupTables: feature,
+                                                messageStore: messageStore) {
+                return message
+            }
+            
+            onMalformedMessage(messageKey: key)
+            
+            return nil
+        }.filter { message in
+            !messagingHelper.isMessageExpired(message: message)
+        }.filter { message in
+            message.messageData.surface == surface
+        }.sorted { message1, message2 in
+            message1.styleData.priority > message2.styleData.priority
+        }
+        
+        /// Populating showables for the surface gives us a way to access all messages at once.
+        /// And showables will always be up to date when the surface calls for a message.
+        showableMessagesForSurface[surface]?.append(contentsOf: messages)
+        
+        /// Take the first triggered message.
+        guard let message = messages.first(where: { message in
+            do {
+                return try messagingHelper.isMessageEligible(message: message, messageHelper: helper)
+            } catch {
+                return false
+            }
+        }) else {
+            return nil
+        }
+        
+        /// If it's a message under experiment, we need to react to whether it's a control or not.
+        if messagingHelper.isMessageUnderExperiment(experimentKey: feature.messageUnderExperiment, message: message) {
+            /// TODO: Report via telemetry
+            FxNimbus.shared.features.messaging.recordExposure()
```

Comment:
@jhugman I understood exposures were supposed to be recorded when we presented a surface that was experimentable on, regardless of whether or not the surface was being shown?

Would it make sense here to record the exposure on FirefoxHomepage itself?

Reply:
> I understood exposures were supposed to be recorded when we presented a surface that was experimentable on, regardless of whether or not the surface was being shown?

For most Nimbus features you're right.

The general principal here is that we want to record an exposure when the user has been exposed to a treatment (or the placebo).

For most features, that's whenever the user is exposed to the surface. The visible part of the surface is the treatment.

For messaging, which serve messages from multiple places, the message is the treatment (or placebo). So we only record an exposure when the treatment message (or the placebo) is shown to the user.